### PR TITLE
feat(webserver): add graphql api update_thread_message

### PR DESCRIPTION
**Pull request by zwpaper**
related: https://github.com/TabbyML/tabby/pull/3334

### CODE REVIEW

File: ee/tabby-schema/src/schema/thread/types.rs [hunk 1]
```diff
@@ -44,6 +44,15 @@ impl NodeType for Message {
     }
 }
 
+#[derive(GraphQLInputObject, Clone)]
+#[graphql(context = Context)]
+pub struct UpdateMessageRequest {
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/thread/types.rs:49]: please follow the practice of other mutation input - `UpdateMessageInput`

File: ee/tabby-schema/src/schema/mod.rs [hunk 2]
```diff
@@ -1167,6 +1167,31 @@ impl Mutation {
...
+        message: thread::UpdateMessageRequest,
+    ) -> Result<thread::Message> {
+        let user = check_user(ctx).await?;
+        let svc = ctx.locator.thread();
+        let Some(thread) = svc.get(&message.thread_id).await? else {
+            return Err(CoreError::NotFound("Thread not found"));
+        };
+
+        user.policy
+            .check_update_thread_persistence(&thread.user_id)?;
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/mod.rs:1181]: create a new policy for this, e.g `check_update_thread_message`

File: ee/tabby-schema/src/schema/mod.rs [hunk 3]
```diff
@@ -1167,6 +1167,31 @@ impl Mutation {
...
+        let messages = svc
+            .list_thread_messages(&message.thread_id, None, None, None, None)
+            .await?;
+
+        messages
+            .iter()
+            .peekable()
+            .find(|m| m.id == message.id)
+            .cloned()
+            .ok_or_else(|| CoreError::NotFound("Message not found"))
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/mod.rs:1192]: adding a `todo!()` to be more explicit.

File: ee/tabby-schema/src/schema/thread/types.rs [hunk 4]
```diff
@@ -44,6 +44,15 @@ impl NodeType for Message {
...
+    pub id: ID,
+    pub thread_id: ID,
+    pub role: Role,
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/thread/types.rs:52]:
```suggestion
```

Role is not modificable

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-schema/src/schema/mod.rs [hunk 5]
```diff
@@ -1167,6 +1167,27 @@ impl Mutation {
...
+            return Err(CoreError::NotFound("Thread not found"));
+        };
+
+        user.policy.check_update_thread_message(&thread.user_id)?;
+
+        let message = svc.get_thread_message(&input.id).await?;
+        if input.content == message.content {
+            return Ok(true);
+        }
+
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/mod.rs:1186]: no need to check equivalent, just update the message regardlessly

File: ee/tabby-webserver/src/service/thread.rs [hunk 6]
```diff
@@ -52,13 +52,29 @@ impl ThreadService for ThreadServiceImpl {
...
             .update_thread_ephemeral(id.as_rowid()?, false)
             .await?;
         Ok(())
     }
 
+    async fn update_thread_message(&self, input: &UpdateMessageInput) -> Result<()> {
+        self.db
+            .append_thread_message_content(input.id.as_rowid()?, &input.content)
+            .await?;
+        Ok(())
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/thread.rs:75]: 1. the db function need to take both thread_id and id to ensure id is part of thread_id.
2. update_thread_message_content, not append.

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-db/src/threads.rs [hunk 7]
```diff
@@ -220,6 +220,24 @@ impl DbConn {
         Ok(())
     }
 
+    pub async fn update_thread_message_content(
+        &self,
+        thread_id: i64,
+        message_id: i64,
```
Comment (wsxiaoys) [inline ee/tabby-db/src/threads.rs:226]: need to verify message_id belongs to thread_id

File: ee/tabby-webserver/src/service/thread.rs [hunk 8]
```diff
@@ -52,13 +52,33 @@ impl ThreadService for ThreadServiceImpl {
...
     }
 
+    async fn get_thread_message(&self, id: &ID) -> Result<Message> {
+        Ok(self
+            .db
+            .get_thread_message(id.as_rowid()?)
+            .await?
+            .ok_or_else(|| CoreError::NotFound("Message not found"))?
+            .try_into()?)
+    }
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/thread.rs:62]: this function is not used

File: ee/tabby-db/src/threads.rs (hunk 7 above)
Comment (zwpaper) [reply ee/tabby-db/src/threads.rs:226]: https://github.com/TabbyML/tabby/pull/3344/files/640aeaad3e2819cbe28e6481d27b52859f65ae74#diff-d0fb876834081b31a6bf3d505dfb7c1567ed9fe5c851ac4fa0387f4437b41856R230 

use condition `thread_id = ? and id = ?`, it should report NOT_FOUND when the message_id not belongs to thread_id

File: .changes/unreleased/Features-20241031-152952.yaml
Comment (wsxiaoys) [inline .changes/unreleased/Features-20241031-152952.yaml:1]: For non-user facing change, no need to document it as a "feature" in changie.

**Review by wsxiaoys — APPROVED** [review]

**Review by liangfung — APPROVED** [review]

**Pull request merged**